h/api/nipsa/search.py

```diff
@@ -0,0 +1,61 @@
+def nipsa_filter(userid=None):
+    """Return an Elasticsearch filter for filtering out NIPSA'd annotations.
+
+    The returned filter is suitable for inserting into an Es query dict.
+    For example:
+
+        query = {
+            "query": {
+                "filtered": {
+                    "filter": nipsa_filter(),
+                    "query": {...}
+                }
+            }
+        }
+
+    :param userid: The ID of a user whose annotations should not be filtered.
+        The returned filtered query won't filter out this user's annotations,
+        even if the annotations have the NIPSA flag.
+    :type userid: unicode
+
+    """
+    # If any one of these "should" clauses is true then the annotation will
+    # get through the filter.
+    should_clauses = [{"not": {"term": {"nipsa": True}}}]
+
+    if userid:
```

Comment:
I think it should be explicit here and I don't think that violates the
style guide, either.

For me, it's just not the place for this function to judge whether an empty
string is a valid username. I know that sounds silly, but we're really
trying to check "was I passed an argument", no? Check that.

On Wed, Jul 22, 2015, 07:58 Sean Hammond <notifications@github.com> wrote:

> In h/api/nipsa/search.py
> <https://github.com/hypothesis/h/pull/2300#discussion_r35222179>:
>
> > +                    "query": {...}
> > +                }
> > +            }
> > +        }
> > +
> > +    :param userid: The ID of a user whose annotations should not be filtered.
> > +        The returned filtered query won't filter out this user's annotations,
> > +        even if the annotations have the NIPSA flag.
> > +    :type userid: unicode
> > +
> > +    """
> > +    # If any one of these "should" clauses is true then the annotation will
> > +    # get through the filter.
> > +    should_clauses = [{"not": {"term": {"nipsa": True}}}]
> > +
> > +    if userid:
>
> So you're saying we assume the value is either None or a valid userid (so
> not the empty string and not falsy). In that case if userid: and if
> userid is not None: are the same behaviour-wise and it comes down to
> style. Our style guide says to use the implicit false *if at all possible*
> and here it is possible to get the right behaviour by using it ... So
> unless we want to introduce a new style rule along the lines of always use is
> None instead of the implicit false, or always use it in case of None as a
> sentinel value, or something.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/2300/files#r35222179>.
>